### PR TITLE
Make unwinding of samples falling inside uprobes-instrumented functions work

### PR DESCRIPTION
**Pull request by dpallotti**
Uretprobes causes unwinding of stack samples falling inside uprobes-instrumented
functions to stop at the first of such functions to be encountered. We solve
this by processing u(ret)probes and stack samples in order and keeping a stack
of collected at the beginning of uprobes-instrumented functions to rebuild such
broken callstacks. See LinuxUprobesUnwindingVisitor.h for more details.

### CODE REVIEW

**Comment by dpallotti** [comment]
> Now that gtest is enabled we should probably start adding tests. Can you please add unit-tests for this?

I don't see the comments on the code anymore but I made changed to have the output of the tests readable.

File: OrbitCore/LinuxPerfEventProcessor.h [hunk 1]
```diff
@@ -15,63 +18,56 @@
...
-  // While processing, we do not touch the events with a timestamp less
-  // than 1/10 sec smaller than the most recent one in the queue.
-  // This way we can ensure, that all events (from different sources)
-  // are processed in the correct order.
-  const uint64_t DELAY_IN_NS = 100000000 /*ns*/;
+  // Do not process events that are more recent than 0.1 seconds. There could be
+  // events coming out of order as they are read from different perf_event_open
+  // ring buffers and this ensure that all events are processed in the correct
+  // order.
+  static constexpr uint64_t DELAY_NS = 100'000'000;
```
[the remark this answers is not in the document]
Comment (dpallotti) [reply OrbitCore/LinuxPerfEventProcessor.h]: Done.

File: OrbitCore/LinuxPerfUtils.h [hunk 2]
```diff
@@ -19,13 +19,21 @@
...
+#include <ctime>
 
 inline int32_t perf_event_open(struct perf_event_attr* attr, pid_t pid,
                                int32_t cpu, int32_t group_fd, uint64_t flags) {
   return syscall(__NR_perf_event_open, attr, pid, cpu, group_fd, flags);
 }
 
 namespace LinuxPerfUtils {
+
+inline uint64_t GetClockRealtime() {
```
[the remark this answers is not in the document]
Comment (dpallotti) [reply OrbitCore/LinuxPerfUtils.h]: Oh wow, I keep getting names wrong.

**Comment by dpallotti** [comment]
> The only thing is that I wonder if there would be a more direct way of reconstructing the callstacks. We're doing a lot of work it seems, bookkeeping, string compares for "[uprobes]", etc. If we knew exactly where to look for the original return addresses, maybe we'd be able to get the callstacks more efficiently. Also, if we get events out of order, it seems this approach breaks down. That being said, this is clearly way better than what we have right now, so I'm all for merging it.

I had tried to look into retrieving the original address but it seems to be stored in kernel's memory and I got stuck, so honestly I'm very happy we found this solution. Especially the priority queue could have quite an impact, but unless we switch to frame-pointer unwinding, DWARF-based unwinding itself also has quite an impact. We should obviously measure it at some point, but for now we're trying to get things to work.
As for getting events out of order, as long as the timestamps are correct it should always be possible to find a solution to make sure events are processed in order.

**Comment by pierricgimmig** [comment]
> I had tried to look into retrieving the original address but it seems to be stored in kernel's memory and I got stuck, so honestly I'm very happy we found this solution.

Ah ok I see, thanks for the added context.  I'm happy too!

File: OrbitCore/LinuxUprobesUnwindingVisitorTests.cpp [hunk 3]
```diff
@@ -1,3 +1,292 @@
...
+  return vector;
+}
+}  // namespace
+
+TEST(UprobesCallstackManager, NoUprobes) {
+  std::vector<unwindstack::FrameData> unwound_cs, real_cs, processed_cs;
+  UprobesCallstackManager callstack_manager{};
+
+  unwound_cs = real_cs = MakeTestCallstack({"main", "alpha", "beta"});
+  processed_cs = callstack_manager.ProcessSampledCallstack(42, unwound_cs);
```
Comment (dimitry-) [inline OrbitCore/LinuxUprobesUnwindingVisitorTests.cpp]: add constexpr pid_t tid = 42 and use it here and below?

File: OrbitCore/LinuxUprobesUnwindingVisitorTests.cpp [hunk 4]
```diff
@@ -1,3 +1,292 @@
...
+  }
+  return vector;
+}
+}  // namespace
+
+TEST(UprobesCallstackManager, NoUprobes) {
+  std::vector<unwindstack::FrameData> unwound_cs, real_cs, processed_cs;
+  UprobesCallstackManager callstack_manager{};
+
+  unwound_cs = real_cs = MakeTestCallstack({"main", "alpha", "beta"});
```
Comment (dimitry-) [inline OrbitCore/LinuxUprobesUnwindingVisitorTests.cpp]: s/real_cs/expected_cs/ ?

**Pull request merged**